### ISSUE TITLE
Blazor browser crashes

### ISSUE THREAD

**Issue by connor4312** [dotnet/aspnetcore#25519]
> Hi, I came here from https://devblogs.microsoft.com/aspnet/blazor-webassembly-3-2-0-preview-3-release-now-available/
> And seems like `pwa-msedge` doesn't work.
> I have latest MS Edge installed ( Version 80.0.361.69 (Official build) (64-bit) ) `"debug.javascript.usePreview": true` and this is my launch.json:
> 
> ```js
> {
>             "name": ".NET Core Debug Blazor Web Assembly in Edge",
>             "type": "pwa-msedge",
>             "request": "launch",
>             "timeout": 30000,
>             "url": "https://localhost:5001",
>             "webRoot": "${workspaceFolder}/Server",
>             "inspectUri": "{wsProtocol}://{url.hostname}:{url.port}/_framework/debug/ws-proxy?browser={browserInspectUri}"
> }
> ```
> 
> When I trying to start debug I see this error:
> ![image](https://user-images.githubusercontent.com/40987940/77764784-42480e80-704e-11ea-99f8-7cf44fbd8c1e.png)
> 
> ```
> Unable to launch browser: "Failed to launch browser!
> [0327/171300.005:ERROR:file_io.cc(89)] ReadExactly: expected 120, observed 40
> 
> TROUBLESHOOTING: https://github.com/GoogleChrome/puppeteer/blob/master/docs/troubleshooting.md
> ```
> 
> Can you help me find out what's wrong?
> 
_Originally posted by @vchirikov in https://github.com/microsoft/vscode-js-debug/issues/405

I've been getting a steady trickle of reports on this from people using Blazor, reported on both Chrome and Edge. @digeff reports that they haven't seen this on VS, and I haven't received reports on this from any non-Blazor users. Is there anything Blazor is doing differently that could cause this? Trees to bark up would also be appreciated 😛

**Comment by Stryxus** [comment]
Side note: The latest version of MS Edge is 85.0.564.44.

And personally I launch the project directly and type in localhost:5000 manually, its much easier, especially with WebAssembly. Well for me at least.

My launchSettings.json:
`{
    "profiles": {
      "SA.Web.Server": {
        "commandName": "Project",
        "launchBrowser": false,
        "inspectUri": "{wsProtocol}://{url.hostname}:{url.port}/_framework/debug/ws-proxy?browser={browserInspectUri}",
        "applicationUrl": "http://localhost:5000",
        "environmentVariables": {
          "ASPNETCORE_ENVIRONMENT": "Development"
        }
      }
    }
  }
`

**Comment by captainsafia** [comment]
@connor4312 The debugging setup in Blazor is definitely a little different than a standard JS debugging experience.

I'm a little surprised at the exception thrown here. It seems to be coming from the Chrome launcher itself, which comes into play before the Blazor app has launched. The entire flow for debugging Blazor WASM in VS Code looks something like this (assuming the user uses `blazorwasm` debug type).

1. User presses F5.
2. VS Code passes the debugging request to our debugging config resolver.
3. The blazorwasm config provider spawns a process running `dotnet run` in the current working directory (or a cwd provided by the user).
4. The `blazorwasm` config provider starts a JS debugging session throw either pwa-msedge or pwa-chrome.
5. The JS debugger sends a request to http://localhost:5001/_framework/debug/ws-proxy/...
6. The Blazor server receives the request from #5 and starts the Mono debug proxy. The JS debugger sends CDP messages to this endpoint.
7. The Chrome browser is started.

You might have some more insights, @connor4312, about the exact ordering of 5-7. Perhaps the fact that the JS debugger communicates with the application server causes some issues here? Although I dunno why that would surface the stack traces we are seeing from Chrome.

**Comment by connor4312** [comment]
It seems like we haven't gotten any trace logs yet, so I've asked in the other issue to collect some. That should help understand the ordering.

**Comment by aventus13** [comment]
Please see attached for the 
[vscode-debugadapter-0.json.gz](https://github.com/dotnet/aspnetcore/files/5164848/vscode-debugadapter-0.json.gz)
trace logs

**Comment by connor4312** [comment]
That appears to be a separate issue, looks like the Blazor proxy is not running:

> "Unable to launch browser: \"Could not open ws://localhost:5001/_framework/debug/ws-proxy?browser=ws%3A%2F%2F127.0.0.1%3A50835%2Fdevtools%2Fbrowser%2F57772b19-838c-45c9-9a60-20ea02b21f62\""

**Comment by aventus13** [comment]
Should I raise a separate issue for it then? Am I missing some setup? How do I run the Blazor proxy?

**Comment by captainsafia** [comment]
@aventus13 You're in our repo now. Can you share the version of Blazor that you were using when you encountered this issue?

Thanks,
Safia

**Comment by vchirikov** [comment]
And it still doesn't work :D

Today I tried:
```
dotnet new blazorwasm --no-restore
```
launch.json:

```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "type": "blazorwasm",
            "name": "Launch and Debug Blazor WebAssembly Application",
            "request": "launch",
            "browser": "edge",
        }
    ]
}
```
`ms-dotnettools.blazorwasm-companion-1.1.0`
dotnet 6.0

![bug](https://user-images.githubusercontent.com/40987940/141658458-f065682a-2243-4594-90b2-ea50a1697968.gif)


cc: @pranavkm @captainsafia

**Comment by captainsafia** [comment]
@vchirikov Can you file another issue? The problem you are reporting is distinct from the one discussed here.

**Comment by vchirikov** [comment]
@captainsafia I found the reason, it's occupied port 9222.  
vs / vscode use the hard-coded value `--remote-debugging-port=9222` and I'm already using the port in another browser.
That's why msedge crashes. By the way, Rider uses `--remote-debugging-port=0` and it worked without any investigations :) I think vs / vscode should check open ports before starting debugging or use a random port.

**Comment by captainsafia** [comment]
> @captainsafia I found the reason, it's occupied port 9222. vs / vscode use the hard-coded value `--remote-debugging-port=9222` and I'm already using the port in another browser. That's why msedge crashes. By the way, Rider uses `--remote-debugging-port=0` and it worked without any investigations :) I think vs / vscode should check open ports before starting debugging or use a random port.

I believe VS already uses a random port and VS Code looks for available ports on the machine to use ([ref](https://github.com/dotnet/razor-tooling/blob/main/src/Razor/src/Microsoft.AspNetCore.Razor.VSCode.BlazorWasmDebuggingExtension/src/getAvailablePort.ts#L8-L26)) so I'm not sure that would explain the issue.

In any case, I'm glad you were able to find a resolution.

**Comment by vchirikov** [comment]
@captainsafia your code doesn't work:
![image](https://user-images.githubusercontent.com/40987940/142150262-5bf95de7-3301-47c6-b6fe-2ad5215b20a4.png)
